File: hashing/counting/c++/longest_substring/longest_substring.cpp
// longest_substring.cpp
#include "longest_substring.hpp"
#include <string>
#include <unordered_map>

int longestSubstring(const std::string &s, const int k)
{
    int left = 0, windowLen = 0;
    std::unordered_map<char, int> counts;
    for (int right = 0; right < s.size(); right++)
    {
        // [] automatically inserts the key with a default value (0 for integers) if it doesn't exist
        char rightChar = s[right];
        counts[rightChar]++;

        // Check if the current window has become invalid (i.e. > k distinct characters)
        while (counts.size() > k)
        {
            // Decrement the counter of the key pointed to by the left pointer
            char leftChar = s[left];
            counts[leftChar]--;
            // Check if the count has reached zero, if so, remove the key
            if (counts[leftChar] == 0)
            {
                counts.erase(leftChar);
            }
            left++;
        }
        windowLen = std::max(windowLen, right - left + 1);
    }

    return windowLen;
}

File: hashing/counting/python3/longest_substring.py
from collections import defaultdict
from string import ascii_lowercase

import numpy as np


def longest_substring(s: str, k: int) -> int:
    """
    Given a string `s` and an integer `k`. Find the length of the longest substring that contains at most `k` distinct characters.

    Parameters
    ----------
    s : str
        A string of lowercase English letters
    k : int
        An integer

    Returns
    -------
    int
        The length of the longest substring that contains at most `k` distinct characters
    """
    counts = defaultdict(int)
    left = window_len = 0
    for right in range(len(s)):
        # Slide the window forward
        counts[s[right]] += 1
        # Check if the window has become invalid (i.e. more than k distinct characters)
        while len(counts) > k:
            # Decrement the count of the character the left pointer points to
            counts[s[left]] -= 1
            # If this count becomes zero, remove the key from the map
            if counts[s[left]] == 0:
                del counts[s[left]]
            # Increment the left pointer
            left += 1
        # Check if the answer should be updated
        window_len = max(window_len, right - left + 1)
    return window_len


def main() -> int:
    letters = list(ascii_lowercase)
    for _ in range(4):
        length = np.random.randint(low=4, high=15)
        s = "".join(np.random.choice(a=letters, size=(length,)))
        k = np.random.randint(low=1, high=length - 2)
        ans = longest_substring(s, k)
        print(
            f"Given the string {s} and k = {k}, the longest substring with {k} distinct character has length {ans}"
        )

    return 0


if __name__ == "__main__":
    main()
